### PR TITLE
[JENKINS-43444] Connect new JNLP Agent should be one-click

### PR DESCRIPTION
**Pull request by batmat**
https://issues.jenkins-ci.org/browse/JENKINS-43444
Not only provide the `java -jar ...` command, but also a `curl` or Windows command to download the 
then run the slave.jar. Plus a nowadays more and more common "Copy" button to save yet a bit more time.

Screenshot of the slightly changed UI (if you click on the `Linux` button, which is the default display when the page shows up):

image

If you click on the `Windows` button:

image

From a technical standpoint, essentially:
* defines 'copiable' class that then enables the auto addition of a copy
  button for marked textareas
* Use of `data-windows` and `data-linux` attributes to store command for respective
  platforms. I could have used a JS var to fill the field when showing the
  page, but I wanted to have something that would still show up in case JS
  is disabled or at least a bit broken (and not an empty textarea in that
  case)

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Maybe `curl ... && java ...` then? Just to make it copy&paster

**Comment by batmat** [comment]
@oleg-nenashev why not, but actually it already works by copy pasting. It will just then execute indeed the first line right away, do you think this is an issue? Maybe so.

**Comment by daniel-beck** [comment]
`&&` is invalid in `fish` and possibly other shells, so we're just going to get complaints from super special snowflakes.

**Comment by daniel-beck** [comment]
This ignores Windows and will just make @jtnord cry.

**Comment by batmat** [comment]
I actually chose curl instead of what for that reason : I think it's more common nowadays to have curl in general (than wget), even on Windows. But now I'm at it, I'm thinking about making something similar to what GitHub has on some articles : the command could be switched with a click + a copy to buffer button.
Also: my thought was also that people on even more special platform would just copy the second line as now.

If James thinks there's a better command for Windows though I can consider using it.

**Comment by oleg-nenashev** [comment]
Actually curl is also not guaranteed to be present on the Unix systems. So maybe you could add a comment like "or just download it using the hyperlink below" after the command sample

**Comment by jtnord** [comment]
Windows would use  powershell. 
`Invoke-WebRequest -Uri $theurl -OutFile slave.jar`

Requires IE so won't work apparently on server core installations.

**Comment by oleg-nenashev** [comment]
So do we want to merge it as is? @batmat I am waiting for the feedback from you regarding CURL

**Comment by batmat** [comment]
Yeah, right. Neither is `java` guaranteed to be present actually ;-). I'm marking this as WIP as I need more time to think about what I could do to make it still usable for the core use case I wanted to achieve (make a common Linux agent JNLP Startup quicker thank right-click, type wget or curl, paste the link, back to jenkins, copy/paste the java -jar ... command)

**Comment by batmat** [comment]
@daniel-beck @oleg-nenashev @jtnord I've played a bit with JS, and here's a screenshot of what I have right now:

image

I put the existing text in a readonly textarea. The copy button directly copies the buffer, like it's becoming more common nowadays. WDYT?

Now, looking into making it switchable easily for Windows agents.

**Comment by oleg-nenashev** [comment]
@batmat Looks good to me. Maybe it would be preferable to have "copy" icons instead of buttons, similarly to GitHub and other such services. Maybe frontend guys could suggest something, CC @kzantow , @recena, @tfennelly

**Comment by batmat** [comment]
Yeah, it was a quick first attempt and I didn't want to lose time looking for icons/glyphs. I had indeed in mind to make it a bit more fancy when other things are OK.

**Comment by batmat** [comment]
@oleg-nenashev @daniel-beck @jtnord WDYT, I updated the PR code + the screenshots above (https://github.com/jenkinsci/jenkins/pull/2775#issue-211816469).

I would like to avoid spending more time on this PR to add an icon or so immediately. 

This looks quite fine to me right now from a usability perspective, and at least I think it does already improve the previous state.

If you think this is acceptable as is, I would like to merge it as is, and consider filing a followup PR to replace the "Copy" button with something like a copy icon, but would prefer not to do right now.

WDYT?

File: core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly]: `textarea.copiable[data-linux]` is suspicious

File: core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly]: Not every Linux contains CURL, And not Every non-Windows system is Linux

Comment (oleg-nenashev) [inline core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly]: My recommendation would be to select a neutral term like POSIX or Unix-alike

**Comment by recena** [comment]
I will try to review the code related with the UI.

**Comment by recena** [comment]
@batman I can do the PR followup for polishing the UI/UX. If you agree.

File: core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly
Comment (batmat) [inline core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly]: Actually, I understand your suspicion, but it's not. The right way to do the selection I had in mind would have been something like `textarea.copiable[data-linux&&data-windows]` but I considered it overkill.

**Comment by batmat** [comment]
@recena that would be awesome! Team work :-). :handshake: :raised_hands:

File: core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly
Comment (batmat) [inline core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly]: Yeah, I agree that from a technical standpoint, it's true. But I think putting POSIX instead of Linux would be less clear for a majority of users. Maybe Unix, yes, but well, it's probably also a minority of agents (pretty sure stats would back that btw).

Comment (oleg-nenashev) [inline core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly]: what would you think about just replacing the text by `CURL`? Seems to be pretty straightforward and there is no "Linux" mention which may piss off other distribution maintainers.

Comment (batmat) [inline core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly]: Sorry, missed your message @oleg-nenashev :disappointed: 
Yeah, I think this may make sense.

But then I leave Windows for the other? Makes sense. Doing it.

Comment (batmat) [inline core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly]: Done on the button/visible text only. But well, now I think I probably should do that too even on the code. Maybe.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Sorry for another 🐛 . "Invoke-WebRequest" is not available in Batch. It is specific to PowerShell only. So I would change it as well

**Comment by oleg-nenashev** [comment]
Maybe a default "With downloaded jar" option is also helpful.

**Comment by batmat** [comment]
Right, will do that later today or during the WE hopefully. NIT: I think you meant :-1: instead of :bug: @oleg-nenashev ;-)

**Comment by oleg-nenashev** [comment]
@batmat Sorry, the bug emoji is so cute in Windows, cannot stop using it. 

capture

**Comment by daniel-beck** [comment]
Would be preferable if somewhere on the page there still remains a regular download link. `curl` is only an improvement if it doesn't replace a generic way to download the file.

**Comment by oleg-nenashev** [comment]
@daniel-beck do you require the change here?

**Comment by daniel-beck** [comment]
@oleg-nenashev Yes, since it got removed from https://github.com/jenkinsci/jenkins/pull/2775/files#diff-63a487bb5047df29065dc7ddbb149176L60

**Comment by oleg-nenashev** [comment]
There will be a conflict with #2875 since I move the code to the plugin

**Comment by daniel-beck** [comment]
Does this need escaping of weird agent names? See e.g. https://issues.jenkins-ci.org/browse/JENKINS-26658 for someone who didn't expect to have to escape `%` in batch.

**Comment by daniel-beck** [comment]
Provided feedback outside the formal review mechanism, so unrequested myself.

**Comment by batmat** [comment]
Closing for now as I don't find time to finish it. Might reopen some day.

**Pull request closed without merging**
